codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parse/JsonParserGenerator.kt

```diff
@@ -0,0 +1,457 @@
+        val members = includedMembers ?: shape.members()
+        if (members.isNotEmpty()) {
+            objectKeyLoop {
+                rustBlock("match key.as_escaped_str()") {
+                    for (member in members) {
+                        rustBlock("${member.wireName().dq()} =>") {
+                            withBlock("builder = builder.${member.setterName()}(", ");") {
+                                deserializeMember(member, false)
+                            }
+                        }
+                    }
+                    rustTemplate("_ => #{skip_value}(tokens)?", *codegenScope)
+                }
+            }
+        }
+    }
+
+    private fun RustWriter.deserializeMember(memberShape: MemberShape, forceOptional: Boolean) {
+        when (val target = model.expectShape(memberShape.target)) {
+            is StringShape -> deserializeString(target)
+            is BooleanShape -> rustTemplate("#{expect_bool_or_null}(tokens.next())?", *codegenScope)
+            is NumberShape -> deserializeNumber(target)
+            is BlobShape -> rustTemplate("#{expect_blob_or_null}(tokens.next())?", *codegenScope)
+            is TimestampShape -> deserializeTimestamp(memberShape)
+            is CollectionShape -> deserializeCollection(target)
+            is MapShape -> deserializeMap(target)
+            is StructureShape -> deserializeStruct(target)
+            is UnionShape -> deserializeUnion(target)
+            is DocumentShape -> rustTemplate("Some(#{expect_document}(tokens)?)", *codegenScope)
+            else -> TODO(target.toString())
+        }
+        val symbol = symbolProvider.toSymbol(memberShape)
+        if (symbol.isBoxed()) {
+            rust(".map(|v| Box::new(v))")
+        }
+        if (!forceOptional && !symbol.isOptional()) {
+            if (symbol.canUseDefault()) {
+                rust(".unwrap_or_default()")
+            } else {
+                val name = escape(memberShape.memberName)
+                rustTemplate(
+                    ".ok_or_else(|| #{Error}::custom(\"value for '$name' cannot be null\"))?",
+                    *codegenScope
+                )
+            }
+        }
+    }
+
+    private fun RustWriter.deserializeString(target: StringShape) {
+        val isEnum = target.hasTrait<EnumTrait>()
+        val stringFn = if (isEnum) "expect_string_or_null" else "expect_unescaped_string_or_null"
+        rustTemplate("#{$stringFn}(tokens.next())?", *codegenScope)
+        if (isEnum) {
+            rust(".map(|s| #T::from(s.as_escaped_str()))", symbolProvider.toSymbol(target))
+        }
+    }
+
+    private fun RustWriter.deserializeNumber(target: NumberShape) {
+        val symbol = symbolProvider.toSymbol(target)
+        rustTemplate("#{expect_number_or_null}(tokens.next())?.map(|v| v.to_#{T}())", "T" to symbol, *codegenScope)
+    }
+
+    private fun RustWriter.deserializeTimestamp(member: MemberShape) {
```

Comment:
I'm still deciding if how I feel about pushing this stuff into smithy-json/smithy-xml. I think it's good? I guess it simplifies codegen a little bit.